### PR TITLE
Replicate portilla simoncelli

### PR DESCRIPTION
**Pull request by kbonnen**
This pull request adds a verified replication of the portilla-simoncelli texture model (https://www.cns.nyu.edu/pub/eero/portilla99-reprint.pdf) to plenoptic.

It will remain a draft while we review/approve the complexTensor updates the the steerable pyramid.  In the meantime, I will also develop a jupyter notebook tutorial example of the replication.

### CODE REVIEW

**Review by lyndond — CHANGES_REQUESTED** [review]
Minor comments. I'm not an expert on the algo so can't offer in-depth critique on that. But I expect on subsequent go-throughs as I gain understanding I'll have some more commments. 

I haven't looked at the ipynbs since I suspect they'll change a lot.

File: examples/testing_portilla_simoncelli.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+to_numpy = lambda x : x.detach().cpu().numpy().squeeze()
+
+image = plt.imread('../data/metal.pgm').astype(float)
+print(image.shape)
+im0 = torch.tensor(image,requires_grad=True,dtype = torch.float32).squeeze().unsqueeze(0).unsqueeze(0)
+x = po.make_basic_stimuli()
+x = x[0,0,:,:].unsqueeze(0).unsqueeze(0)
+print(x.shape)
+model = po.simulate.Portilla_Simoncelli(x.shape[-2:],n_scales=4,n_orientations=4,Na=9)
+y = model.forward(x)
```
[the remark this answers is not in the document]
Comment (billbrod) [reply examples/testing_portilla_simoncelli.py:22]: I'd add on:
- if it's for personal testing, should be removed
- else, should be put either into `tests` or turned into a notebook, with plenty of text

**Review by billbrod — CHANGES_REQUESTED** [review]
Most important thing: I think there's a bug in how the `scales` arg to the forward method works, because it looks like the model returns the same output no matter what you set `scales` equal to (in which case, I'm not sure how coarse-to-fine synthesis is working). 

I'm going to push a commit in a little bit cleaning up some of `portilla_simoncelli.py` (largely removing unnecessary white space, and stuff like that), so double-check that I didn't change anything important. I'll then push another one making use of the osf download as fixtures.

Additional comments:

- What's going on with the two notebooks? Should they be removed (in favor of the one that @kbonnen is working on now, in a new branch)? If not, they need a lot more text.
- And same with the script in `examples` -- that should be deleted, right?

**Comment by billbrod** [comment]
Oh the other thing I wanted to bring up -- I thought we had decided that models shouldn't keep a `self.representation` lying around, so that they're more lightweight. Did I remember that correctly? If so, that will require a bit more work here, but it looks possible to me.

File: tests/test_models.py [hunk 2]
```diff
@@ -139,6 +144,137 @@ def test_grad(self, basic_stim):
         assert y[0].requires_grad
 
 
+class TestPortillaSimoncelli(object):
+    @pytest.mark.parametrize("n_scales", [1, 2, 3, 4])
+    @pytest.mark.parametrize("n_orientations", [2, 3, 4])
+    @pytest.mark.parametrize("spatial_corr_width", [3, 5, 7, 9])
+    @pytest.mark.parametrize("im_shape", [(256, 256)])
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply tests/test_models.py:166]: I used to test multiple values.  I will remove it now.

File: examples/testing_portilla_simoncelli.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import plenoptic as po
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply examples/testing_portilla_simoncelli.py]: I have removed this script

File: examples/testing_portilla_simoncelli.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbonnen) [reply examples/testing_portilla_simoncelli.py]: This was for personal testing.  It got turned into a set of tests already and I forgot to remove it.

File: tests/test_models.py [hunk 4]
```diff
@@ -139,6 +144,137 @@ def test_grad(self, basic_stim):
...
+    ## tests for whether output matches the original matlab output.  This implicitly tests that Portilla_simoncelli.forward() returns an object of the correct size.
+    @pytest.mark.parametrize("n_scales", [1, 2, 3, 4])
+    @pytest.mark.parametrize("n_orientations", [2, 3, 4])
+    @pytest.mark.parametrize("spatial_corr_width", [3, 5, 7, 9])
+    @pytest.mark.parametrize("im_shape", [(256, 256)])
+    @pytest.mark.parametrize("im", ["curie", "einstein", "metal", "nuts"])
+    def test_ps_torch_v_matlab(
+        self, n_scales, n_orientations, spatial_corr_width, im_shape, im
+    ):
+        path = osf_download("portilla_simoncelli_matlab_test_vectors.tar.gz")
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply tests/test_models.py]: BIlly seems to have completed converting the osf_download calls into fixtures so I'm going to mark this as resolved.

File: plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py [hunk 5]
```diff
@@ -226,6 +226,37 @@ def __init__(self, image_shape, height='auto', order=3, twidth=1, is_complex=Fal
 
         # reasonable default dtype
         self = self.to(torch.float32)
+        
+        # initialize empty coefficient dictionary
+        self.pyr_coeffs = OrderedDict()
+        
+        
+    @classmethod
+    def vector_to_pyramid(cls, vector, imShape, Nsc, Nor):
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py]: This was removed.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 6]
```diff
@@ -0,0 +1,1072 @@
...
+from collections import OrderedDict
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+from ...tools.display import clean_up_axes, update_stem, clean_stem_plot
+
+
+class PortillaSimoncelli(nn.Module):
+    r"""Model for measuring statistics originally proposed in [1] for synthesis.
+
+    Currently we do not: support batch measurement of images.
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:17]: I have added exception handling.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 7]
```diff
@@ -0,0 +1,1072 @@
...
+
+    Parameters
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:19]: The notebook that explains this isn't going in until my next pull request.  I have added a couple of lines to this file to give more context.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 8]
```diff
@@ -0,0 +1,1072 @@
...
+        """
+
+        for this_scale in range(0, self.n_scales):
+            band_num_el = self.real_pyr_coeffs[(this_scale, 0)].numel()
+            if this_scale < self.n_scales - 1:
+                next_scale_mag = torch.empty((band_num_el, self.n_orientations))
+                next_scale_real = torch.empty((band_num_el, self.n_orientations * 2))
+
+                for nor in range(0, self.n_orientations):
+                    # upsampled = PortillaSimoncelli.expand(self.pyr_coeffs[(this_scale + 1, nor)].squeeze(), 2) / 4.0
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:633]: Commented code is removed and a more appropriate explanation has been added.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 9]
```diff
@@ -0,0 +1,1072 @@
...
+
+        if self.use_true_correlations:
+            ac = ac / vari
+
+        return ac, vari
+
+    def compute_skew_kurtosis(self, ch, vari):
+        r"""Computes the skew and kurtosis of ch given the ratio of its
+        variance (vari) and the pixel variance of the original image are
+        above a certain threshold.  If the ratio does not meet that threshold
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py]: I can't give a confident/ meaningful answer to this question that I feel comfortable adding into the comments.  That variance lives in the denominator of the skew calculation so if it is very very small it would blow up the skew.  I'm guessing that would lead to unstable synthesis.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 10]
```diff
@@ -0,0 +1,1072 @@
...
+        """
+
+        # implementation is only for real components
+        if mu is None:
+            mu = X.mean()
+        if var is None:
+            var = X.var()
+        return torch.mean(torch.abs(X - mu).pow(4)) / (var.pow(2))
+
+    def plot_representation(
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:909]: Docstring added

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 11]
```diff
@@ -0,0 +1,1072 @@
...
+            if isinstance(v, OrderedDict):
+                ax = clean_stem_plot(list(v.values()), ax, k, ylim=ylim)
+            else:
+                ax = clean_stem_plot(v.flatten().detach().numpy(), ax, k, ylim=ylim)
+
+            axes.append(ax)
+
+        return fig, axes
+
+    def _representation_for_plotting(self, rep, batch_idx=0):
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:996]: Added.

**Comment by kbonnen** [comment]
> Most important thing: I think there's a bug in how the `scales` arg to the forward method works, because it looks like the model returns the same output no matter what you set `scales` equal to (in which case, I'm not sure how coarse-to-fine synthesis is working).
> 

You fixed this!! Thanks.

> I'm going to push a commit in a little bit cleaning up some of `portilla_simoncelli.py` (largely removing unnecessary white space, and stuff like that), so double-check that I didn't change anything important. I'll then push another one making use of the osf download as fixtures.
> 


> Additional comments:
> 
>     * What's going on with the two notebooks? Should they be removed (in favor of the one that @kbonnen is working on now, in a new branch)? If not, they need a lot more text.

These have been removed.

> 
>     * And same with the script in `examples` -- that should be deleted, right?

This has been removed

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 12]
```diff
@@ -0,0 +1,1072 @@
...
+    def forward(self, image, scales=None):
+        r"""Generate Texture Statistics representation of an image (see reference [1])
+
+        Parameters
+        ----------
+        image : torch.tensor
+            A tensor containing the image to analyze. We want to operate
+            on this in the pytorch-y way, so we want it to be 4d (batch,
+            channel, height, width). If it has fewer than 4 dimensions,
+            we will unsqueeze it until its 4d
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:198]: opened a new issue for this

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 13]
```diff
@@ -0,0 +1,1072 @@
...
+
+        Returns
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:200]: pushing a fix for this

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 14]
```diff
@@ -0,0 +1,1072 @@
...
+            ind = torch.LongTensor(
+                [
+                    i
+                    for i, s in enumerate(self.representation_scales)
+                    if s in self.scales
+                ]
+            ).to(device)
+            return representation_vector.index_select(0, ind)
+
+        return representation_vector.unsqueeze(0).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:333]: pushing a fix for this

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 15]
```diff
@@ -0,0 +1,1072 @@
...
+
+    def convert_to_vector(self):
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:335]: issue #142

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 16]
```diff
@@ -0,0 +1,1072 @@
...
+        =======
+        representation_vector: torch.tensor
+            A flattened tensor (1d) containing the measured representation statistics.
+
+        """
+        device = image.device
+        while image.ndimension() < 4:
+            image = image.unsqueeze(0)
+
+        # self.pyr.to(image.dtype)
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py]: This was deleted.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 17]
```diff
@@ -0,0 +1,1072 @@
...
+        self.pyr_coeffs = self.pyr.forward(image)
+        self.representation = OrderedDict()
+
+        ### SECTION 1 (STATISTIC: pixel_statistics) ##################
+        #  Calculate pixel statistics (mean, variance, skew, kurtosis, min, max).
+        self.representation["pixel_statistics"] = OrderedDict()
+        self.representation["pixel_statistics"]["mean"] = torch.mean(image)
+        self.representation["pixel_statistics"]["var"] = torch.var(image)
+        self.representation["pixel_statistics"]["skew"] = PortillaSimoncelli.skew(
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:223]: I have transitioned all of the static methods (skew, kurtosis, and expand) from PortillaSimoncelli to self.__class__.method

On whether or not they should be external general methods... They used to be.  Then I was told to bring them into the class.  Unless there's a strong reason to pull them out, I'm going to leave them where they're at.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 18]
```diff
@@ -0,0 +1,1072 @@
...
+        # and real_pyr_coeffs, which contain the magnitude of the
+        # pyramid coefficients and the real part of the pyramid
+        # coefficients respectively.
+        self.representation["magnitude_means"] = self._calculate_magnitude_means()
+
+        ### SECTION 3 (STATISTICS: auto_correlation_magnitude,
+        #                          skew_reconstructed,
+        #                          kurtosis_reconstructed,
+        #                          auto_correlation_reconstructed) #####
+        #
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:245]: I've updated this comment to explain why they're all computed together (see below as well) and I'm not sure what you mean when you say that it's not right above the place where the stats are computed it.  It would break with the consistency of the commenting style to move them to the method calculate_autocorrelation_skew_kurtosis()  so I'm leaving them where they are above the intialization of the variables which is followed by the method that calculates those values.

[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:245]: I also think that when I change PortillaSimoncelli so that it doesn't contain self.representation, the clarity around where the stats are calculated will be improved.  This will occur in a separate, subsequent pull request.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 19]
```diff
@@ -0,0 +1,1072 @@
...
+        )
+
+        representation_vector = self.convert_to_vector()
+
+        if scales is not None:
+            ind = torch.LongTensor(
+                [
+                    i
+                    for i, s in enumerate(self.representation_scales)
+                    if s in self.scales
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py]: This bug has been fixed.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 20]
```diff
@@ -0,0 +1,1072 @@
...
+        for (key, val) in self.magnitude_pyr_coeffs.items():
+            magnitude_means[key] = torch.mean(val)
+            self.magnitude_pyr_coeffs[key] = (
+                self.magnitude_pyr_coeffs[key] - magnitude_means[key]
+            )  # subtract mean of magnitude
+
+        return magnitude_means
+
+    def expand(im, mult):
+        r"""Resize an image (im) by a multiplier (mult).
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:495]: I have added @staticmethod decorators to these three methods.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 21]
```diff
@@ -0,0 +1,1072 @@
...
+        rep["cross_scale_correlation_real"] = vec[n_filled : (n_filled + nn)].unflatten(
+            0, (2 * self.n_orientations, max(2 * self.n_orientations, 5), self.n_scales)
+        )
+        n_filled += nn
+
+        # var_highpass_residual
+        rep["var_highpass_residual"] = vec[n_filled]
+        n_filled += 1
+
+        return rep
```
[the remark this answers is not in the document]
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:454]: Good catch.  This definitely needs a test.

**Review by billbrod — APPROVED** [review]
Looks good to me

**Comment by billbrod** [comment]
A note on something that Kate and I found: our PS texture metamer synthesis is broken by the changes in torch 1.10 (that is, 1.8 and 1.9 give the same result, but 1.10 gives a different one). After looking into it for a while with Nikhil, we're pretty sure it's because they changed the sub-gradient for `torch.min` and `torch.max` (nothing else seems possibly relevant). The forward pass of `PortillaSimoncelli` is unaffected, so it has to be something to do with autograd. The synthesis refactor (#136) also changed the outputs of metamer synthesis. And these two changes appeared to interact in some way -- the differences get larger when using the refactored `Metamer` with torch 1.10 than with either change alone. The arguments required in order to get good-looking synthesis, as used in our test, changed as the result of this, but they still appear to be good enough for us.

**Review by lyndond — APPROVED** [review]
Looks good to me
